tiledb/sm/buffer/buffer.cc

```diff
@@ -42,45 +42,119 @@ using namespace tiledb::common;
 namespace tiledb {
 namespace sm {
 
-/* ****************************** */
-/*             MACROS             */
-/* ****************************** */
+/* ************** */
+/*   BufferBase   */
+/* ************** */
+BufferBase::BufferBase()
+    : data_(nullptr)
+    , size_(0)
+    , offset_(0){};
+
+BufferBase::BufferBase(void* data, const uint64_t size)
+    : data_(data)
+    , size_(size)
+    , offset_(0){};
+
+/*
+ * const_cast is safe here because BufferBase methods do not modify data.
+ */
+BufferBase::BufferBase(const void* data, const uint64_t size)
+    : data_(const_cast<void*>(data))
+    , size_(size)
+    , offset_(0){};
+
+uint64_t BufferBase::size() const {
+  return size_;
+}
 
-#define MAX(a, b) ((a) > (b) ? (a) : (b))
+void* BufferBase::nonconst_data() const {
+  return data_;
+}
+
+/*
+ * const_cast is safe here because BufferBase methods do not modify data.
+ */
+const void* BufferBase::const_data() const {
+  return const_cast<const void*>(data_);
+}
+
+void* BufferBase::nonconst_unread_data() const {
+  if (data_ == nullptr) {
+    return nullptr;
+  }
+  // Cast to byte type because offset is measured in bytes
+  return static_cast<int8_t*>(data_) + offset_;
+}
+
+const void* BufferBase::const_unread_data() const {
+  return const_cast<const void*>(nonconst_unread_data());
+}
+
+uint64_t BufferBase::offset() const {
+  return offset_;
+}
+
+void BufferBase::reset_offset() {
+  offset_ = 0;
+}
+
+void BufferBase::set_offset(const uint64_t offset) {
+  // If the offset is invalid we ignore it.
+  // Arguably we should throw an exception, but that's too disruptive for now.
```

Comment:
Do we have a sense of how often this happens? What would happen before, maybe we expect to be able to set it past the end, then error out on read?

Reply:
In the scenario you sketch, the effect of this change would be for a fault to be detected earlier, during `set_offset`, and not later at `read` or `write`. Given the way most of the code is simply returns on error, it shouldn't affect anything outside the function body where the defective call resides.

The narrow problem for this PR is that a correct fix for the defect requires maintaining the invariant "offset <= size". How we do that is not important insofar as the issues of this PR are concerned. The arithmetic for detecting a read error, in order to avoid overflow, requires it. It would be possible to make a more complicated overflow check during read, but that doesn't seem worth it. 

The "for now" comment really means "during the development of this PR". I wanted to see if CI passes with a minimal modification to its behavior so that it wasn't conflating a fix for this with detection of other defects. That's done, so the next version will have an assertion on the condition rather than returning on it. If that halts a function that would otherwise return an error, we can fix them.

If it's happening now, it would be seeing nonsense on read and clobbering something outside its allocation on write. If we've got such defective calls to this in our code, they might not be faulting in our unit tests, or any faults may not be escalating into errors. Then again, ASAN would detect a write fault easily, but we're not regularly testing with that quite yet.

Reply:
Assert would be a great start!

Reply:
Ok. Using `assert` does not cause all hell to break loose. We're going with that.